### ISSUE TITLE
openblas: segmentation fault with Intel 16.0.3

### ISSUE THREAD

**Issue by weijianwen** [spack/spack#3036]
Spack failed to patch openblas 0.2.15. The same error messages arise when I tried to patch it manually.

This patch is generated by git. I am not sure if incompatability between `diff` and `git diff` lead to this issue. Similar issue happen to OpenMPI 2.0.1. In #2962  , I replaced the original git-generated patch file with one generated locally with `diff`.

```
% spack install openblas@0.2.15 %intel@16.0.3
==> Installing openblas
==> Using cached archive: /tmp/rpm/spack_source_cache/openblas/openblas-0.2.15.tar.gz
==> Already staged openblas-0.2.15-m6w5vfvkcstbmw6pcgwb6w6ovsljsuge in /home/rpm/spack/var/spack/stage/openblas-0.2.15-m6w5vfvkcstbmw6pcgwb6w6ovsljsuge
==> Patching failed last time. Restaging.
==> Staging archive: /home/rpm/spack/var/spack/stage/openblas-0.2.15-m6w5vfvkcstbmw6pcgwb6w6ovsljsuge/v0.2.15.tar.gz
==> Applied patch openblas_icc.patch
1 out of 1 hunk FAILED -- saving rejects to file Makefile.system.rej
1 out of 2 hunks FAILED -- saving rejects to file getarch.c.rej
==> Patch make.patch failed.
==> Error: ProcessError: Command exited with status 1:
    '/usr/bin/patch' '-s' '-p' '1' '-i' '/home/rpm/spack/var/spack/repos/builtin/packages/openblas/make.patch'
/home/rpm/spack/lib/spack/spack/package.py:1020, in do_patch:
     967      def do_patch(self):
     968          """Calls do_stage(), then applied patches to the expanded tarball if they
     969             haven't been applied already."""
     970          if not self.spec.concrete:
     971              raise ValueError("Can only patch concrete packages.")
     972
     973          # Kick off the stage first.
     974          self.do_stage()
     975
     976          # Package can add its own patch function.
     977          has_patch_fun = hasattr(self, 'patch') and callable(self.patch)
     978
     979          # If there are no patches, note it.
     980          if not self.patches and not has_patch_fun:
     981              tty.msg("No patches needed for %s" % self.name)
     982              return
     983
     984          # Construct paths to special files in the archive dir used to
     985          # keep track of whether patches were successfully applied.
     986          archive_dir = self.stage.source_path
     987          good_file = join_path(archive_dir, '.spack_patched')
     988          no_patches_file = join_path(archive_dir, '.spack_no_patches')
     989          bad_file = join_path(archive_dir, '.spack_patch_failed')
     990
     991          # If we encounter an archive that failed to patch, restage it
     992          # so that we can apply all the patches again.
     993          if os.path.isfile(bad_file):
     994              tty.msg("Patching failed last time. Restaging.")
     995              self.stage.restage()
     996                                                                                                   [22/156]
     997          self.stage.chdir_to_source()
     998
     999          # If this file exists, then we already applied all the patches.
     1000         if os.path.isfile(good_file):
     1001             tty.msg("Already patched %s" % self.name)
     1002             return
     1003         elif os.path.isfile(no_patches_file):
     1004             tty.msg("No patches needed for %s" % self.name)
     1005             return
     1006
     1007         # Apply all the patches for specs that match this one
     1008         patched = False
     1009         for spec, patch_list in self.patches.items():
     1010             if self.spec.satisfies(spec):
     1011                 for patch in patch_list:
     1012                     try:
     1013                         patch.apply(self.stage)
     1014                         tty.msg('Applied patch %s' % patch.path_or_url)
     1015                         patched = True
     1016                     except:
     1017                         # Touch bad file if anything goes wrong.
     1018                         tty.msg('Patch %s failed.' % patch.path_or_url)
     1019                         touch(bad_file)
  >> 1020                         raise
     1021
     1022         if has_patch_fun:
     1023             try:
     1024                 self.patch()
     1025                 tty.msg("Ran patch() for %s" % self.name)
     1021
     1022         if has_patch_fun:
     1023             try:
     1024                 self.patch()
     1025                 tty.msg("Ran patch() for %s" % self.name)
     1026                 patched = True
     1027             except:
     1028                 tty.msg("patch() function failed for %s" % self.name)
     1029                 touch(bad_file)
     1030                 raise
     1031
     1032         # Get rid of any old failed file -- patches have either succeeded
     1033         # or are not needed.  This is mostly defensive -- it's needed
     1034         # if the restage() method doesn't clean *everything* (e.g., for a repo)
     1035         if os.path.isfile(bad_file):
     1036             os.remove(bad_file)
     1037
     1038         # touch good or no patches file so that we skip next time.
     1039         if patched:
     1040             touch(good_file)
     1041         else:
     1042             touch(no_patches_file)
```

```
$ tar xzvpf /tmp/rpm/spack_source_cache/openblas/openblas-0.2.15.tar.gz
$ cd OpenBLAS-0.2.15
$ patch -s -p 1 -i /home/rpm/spack/var/spack/repos/builtin/packages/openblas/make.patch
1 out of 1 hunk FAILED -- saving rejects to file Makefile.system.rej
1 out of 2 hunks FAILED -- saving rejects to file getarch.c.rej
```

```
$ head -2 /home/rpm/spack/var/spack/repos/builtin/packages/openblas/make.patch
diff --git a/Makefile.system b/Makefile.system
index b89f60e..2dbdad0 100644
```

**Comment by weijianwen** [comment]
Turns out that `make.patch` doesn't fit openblas-0.2.15 because corresponding lines doesn't exist in that version. Limiting patches to openblas >= 0.2.16 passes the patching phase but new issues arise in the post-installation phase.

```
$ spack install openblas@0.2.15 %intel@16.0.3
==> Installing openblas
==> Using cached archive: /tmp/rpm/spack_source_cache/openblas/openblas-0.2.15.tar.gz
==> Staging archive: /home/rpm/spack/var/spack/stage/openblas-0.2.15-m6w5vfvkcstbmw6pcgwb6w6ovsljsuge/v0.2.15.tar.gz
==> Created stage in /home/rpm/spack/var/spack/stage/openblas-0.2.15-m6w5vfvkcstbmw6pcgwb6w6ovsljsuge
==> Applied patch openblas_icc.patch
==> Building openblas [MakefilePackage]
==> Executing phase : 'edit'
==> Executing phase : 'build'
==> Executing phase : 'install'
==> Error: ProcessError: Command exited with status -11:
    './check'
/home/rpm/spack/var/spack/repos/builtin/packages/openblas/package.py:160, in check_install:
     141      @run_after('install')
     142      def check_install(self):
     143          spec = self.spec
     144          # Openblas may pass its own test but still fail to compile Lapack
     145          # symbols. To make sure we get working Blas and Lapack, do a small
     146          # test.
     147          source_file = join_path(os.path.dirname(self.module.__file__),
     148                                  'test_cblas_dgemm.c')
     149          blessed_file = join_path(os.path.dirname(self.module.__file__),
     150                                   'test_cblas_dgemm.output')
     151
     152          include_flags = ["-I%s" % join_path(spec.prefix, "include")]
     153          link_flags = self.lapack_libs.ld_flags.split()
     154          if self.compiler.name == 'intel':
     155              link_flags.extend(["-lifcore"])
     156          link_flags.extend(["-lpthread"])
     157          if '+openmp' in spec:
     158              link_flags.extend([self.compiler.openmp_flag])
     159
  >> 160          output = compile_c_and_execute(source_file, include_flags, link_flags)
     161          compare_output_file(output, blessed_file)

See build log for details:
  /tmp/rpm/pytest-of-rpm/pytest-4/test_fetch0/tmp/spack-stage/spack-stage-1S4Vxb/OpenBLAS-0.2.15/spack-build.out

$ Generating OpenBLASConfig.cmake in /lustre/spack/sandybridge/linux-centos7-x86_64/intel-16.0.3/openblas-0.2.15-m6w5vfvkcstbmw6pcgwb6w6ovsljsuge/lib/cmake/openblas
Generating OpenBLASConfigVersion.cmake in /lustre/spack/sandybridge/linux-centos7-x86_64/intel-16.0.3/openblas-0.2.15-m6w5vfvkcstbmw6pcgwb6w6ovsljsuge/lib/cmake/openblas
Install OK!
make[1]: Leaving directory `/tmp/rpm/pytest-of-rpm/pytest-4/test_fetch0/tmp/spack-stage/spack-stage-1S4Vxb/OpenBLAS-0.2.15'
==> '/home/rpm/spack/lib/spack/env/cc' '-c' '-I/lustre/spack/sandybridge/linux-centos7-x86_64/intel-16.0.3/openblas-0.2.15-m6w5vfvkcstbmw6pcgwb6w6ovsljsuge/include' '/home/rpm/spack/var/spack/repos/builtin/packages/openblas/test_cblas_dgemm.c'
==> '/home/rpm/spack/lib/spack/env/cc' '-o' 'check' 'test_cblas_dgemm.o' '-L/lustre/spack/sandybridge/linux-centos7-x86_64/intel-16.0.3/openblas-0.2.15-m6w5vfvkcstbmw6pcgwb6w6ovsljsuge/lib' '-lopenblas' '-lifcore' '-lpthread' '-qopenmp'
==> './check'
```

The wierd this is that, even log says 'Install OK' I cannot find installed openblas files in `/lustre/spack/sandybridge/linux-centos7-x86_64/intel-16.0.3/openblas-0.2.15-m6w5vfvkcstbmw6pcgwb6w6ovsljsuge/l`.

**Comment by adamjstewart** [comment]
> Limiting patches to openblas >= 0.2.16 passes the patching phase

Can you please submit a PR with this? Looks like @eschnett didn't check older versions to make sure his patch worked in #886.

> The wierd this is that, even log says 'Install OK' I cannot find installed openblas files in `/lustre/spack/sandybridge/linux-centos7-x86_64/intel-16.0.3/openblas-0.2.15-m6w5vfvkcstbmw6pcgwb6w6ovsljsuge/l`.

It looks like you crashed during `check_install`. This is a function that we wrote to triple check the installation. The installation went okay, but our checks failed, so Spack removed the installation because there is a chance it doesn't work. Otherwise, you would end up in a state where the package is installed but installed incorrectly. 

@davydden It looks like you wrote this in #829. Any idea why it wouldn't be working?

**Comment by davydden** [comment]
> It looks like you wrote this in #829. Any idea why it wouldn't be working?

should be in logs if one keeps the stage. Like compiler/linker error or something along these lines.

**Comment by davydden** [comment]
@adamjstewart @weijianwen just realized that you have those logs, but it looks like a run-time error as `./check` does not print out anything.

**Comment by weijianwen** [comment]
I tried the following steps to verify the possibly linking issue against intel compiler 16.0.

1. Comment the post-installation phase so `openblas@0.2.15 %intel@16.0.3` can be insalled.
2. Build and run `check`.

`./check` gives segment faults this time. That's why OpenBLAS 0.2.15 fails to pass the post-installation check. Logs are attached as follows.

The embarrassing thing is that: for openblas=0.2.1, post-installation `check` fails; for openblas>=0.2.16, it fails in internal tests. The later one remains unsolved in the upstream repo. I plan to:

* Check if proper compiler or linker flags are applied when building OpenBLAS 0.2.15 against Intel 16.0.
* Check if there is patch for openblas>=02.16.

If it fails, I should tell users not to bother compiling openblas against Intel compiler. Just use MKL instead.

```
$ spack install openblas@0.2.15 %intel@16.0.3
$ cd /tmp
$ module load icc/16.0
$ icc '-c' '-I/lustre/spack/sandybridge/linux-centos7-x86_64/intel-16.0.3/openblas-0.2.15-m6w5vfvkcstbmw6pcgwb6w6ovsljsuge/include' '/home/rpm/spack/var/spack/repos/builtin/packages/openblas/test_cblas_dgemm.c'
$ icc '-o' 'check' 'test_cblas_dgemm.o' '-L/lustre/spack/sandybridge/linux-centos7-x86_64/intel-16.0.3/openblas-0.2.15-m6w5vfvkcstbmw6pcgwb6w6ovsljsuge/lib' '-lopenblas' '-lifcore' '-lpthread' '-qopenmp'
$ export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/lustre/spack/sandybridge/linux-centos7-x86_64/intel-16.0.3/openblas-0.2.15-m6w5vfvkcstbmw6pcgwb6w6ovsljsuge/lib
$ ldd ./check | grep openblas
ldd ./check
        libopenblas.so.0 => /lustre/spack/sandybridge/linux-centos7-x86_64/intel-16.0.3/openblas-0.2.15-m6w5vfvkcstbmw6pcgwb6w6ovsljsuge/lib/libopenblas.so.0 (0x00007f279cd06000)
$ ./check
[1]    132617 segmentation fault (core dumped)  ./check
```

**Comment by weijianwen** [comment]
I retried the verification steps above in `openblas@0.2.15 %intel@15.0.3`. It works.👍👍👍 So this should be an issue related to intel@16.0 . I'll check if there is any progress on adapting OpenBLAS for Intel compiler 16.0.

**Comment by davydden** [comment]
> ./check gives segment faults this time. That's why OpenBLAS 0.2.15 fails to pass the post-installation check.

so it's actually good that we had this test :smile:

I suggest you to report segmentation fault in https://github.com/xianyi/OpenBLAS
you may want to run it in debugger so that you get more info on where segfault happens.

**Comment by davydden** [comment]
@weijianwen 

> If it fails, I should tell users not to bother compiling openblas against Intel compiler.

would you mind adding an Assert to throw an error if somebody try to use `intel@16` with `openblas`, until this issue is solved (preferably upstream)?

**Comment by weijianwen** [comment]
Sure. Digging into the segmentn fault and warn the users on intel@16 are actions I can take.

**Comment by weijianwen** [comment]
I am trying to find the root causes of build failures when buidling openblas against Intel compilers. 

openblas@0.2.15 %intel@15.0.2: Build succeeds.
openblas@0.2.16 %intel@15.0.2: Linking error: undefined reference to zerbla_.
openblas@0.2.17 %intel@15.0.2: Linking error: undefined reference to zerbla_. 
openblas@0.2.18 %intel@15.0.2: Linking error: undefined reference to zerbla_. 
openblas@0.2.19 %intel@15.0.2: Build succeeds. 

openblas@0.2.15 %intel@16.0.3: Post-installation check fails due to segment faults.
openblas@0.2.16 %intel@16.0.3: Linking error: undefined reference to zerbla_.
openblas@0.2.17 %intel@16.0.3: Linking error: undefined reference to zerbla_. 
openblas@0.2.18 %intel@16.0.3: Linking error: undefined reference to zerbla_. 
openblas@0.2.19 %intel@16.0.3: Post-installation check fails due to segment faults.

Good news is that there seems to be a patch for zerbla issue. See https://github.com/xianyi/OpenBLAS/issues/910 . This issue, appearing from v0.2.16 to v 0.2.8, is finally solved in v0.2.19.

**Comment by alalazo** [comment]
Closing, as this issue seems outdated. Please reopen if you think this needs further effort.
